### PR TITLE
Ports RBMK reactor

### PR DESCRIPTION
**Pull request by ghost**
# Document the changes in your pull request

Ports RBMK reactor

https://github.com/Quasar-13/Quasar-13/pull/92

I hope this shit will work and i wouldn't close it as soon as i will open this PR

# Changelog

:cl:  
rscadd: Ports RBMK reactor
/:cl:

### CODE REVIEW

File: code/modules/cargo/packs.dm [hunk 1]
```diff
@@ -2676,3 +2676,39 @@
 					/obj/item/wallframe/telescreen/preset,
 					/obj/item/wallframe/telescreen/preset)
 	crate_name = "telescreen crate"
+
+/datum/supply_pack/engine/bananium_fuel_rod
+	name = "Bananium Fuel Rod crate"
+	desc = "Two fuel rods designed to utilize and multiply bananium in a reactor, requires CE access to open. Caution: Radioactive"
+	cost = CARGO_CRATE_VALUE*20
```
Comment (redmoogle) [inline code/modules/cargo/packs.dm]:
```suggestion
	cost = 10000
```

File: code/modules/cargo/packs.dm [hunk 2]
```diff
@@ -2676,3 +2676,39 @@
...
+					/obj/item/fuel_rod/material/bananium)
+	crate_name = "Bluespace Crystal Fuel Rod crate"
+	crate_type = /obj/structure/closet/crate/secure/engineering
+	dangerous = TRUE
+	contraband = TRUE
+
+/datum/supply_pack/engine/fuel_rod
+	name = "Uranium Fuel Rod crate"
+	desc = "Two additional fuel rods for use in a reactor, requires CE access to open. Caution: Radioactive"
+	cost = CARGO_CRATE_VALUE*15
```
Comment (redmoogle) [inline code/modules/cargo/packs.dm]:
```suggestion
	cost = 5000
```
Feel free to use whatever values you want though

File: code/modules/cargo/packs.dm [hunk 3]
```diff
@@ -2676,3 +2676,39 @@
...
+	contains = list(/obj/item/fuel_rod,
+					/obj/item/fuel_rod)
+	crate_name = "Uranium-235 Fuel Rod crate"
+	crate_type = /obj/structure/closet/crate/secure/engineering
+	dangerous = TRUE
+
+/datum/supply_pack/engine/reactor
+	name = "RMBK Nuclear Reactor Kit" // (not) a toy
+	desc = "Contains a reactor beacon and 3 reactor consoles. Uranium rods not included."
+	cost = CARGO_CRATE_VALUE*60
```
Comment (redmoogle) [inline code/modules/cargo/packs.dm]:
```suggestion
	cost = 30000
```

File: code/modules/power/reactor/rbmk.dm [hunk 4]
```diff
@@ -0,0 +1,1031 @@
...
+		color = null
+	else
+		if(has_fuel())
+			temperature += temperature / 500 //This isn't really harmful early game, but when your reactor is up to full power, this can get out of hand quite quickly.
+			vessel_integrity -= temperature / 200 //Think fast loser.
+			take_damage(10) //Just for the sound effect, to let you know you've fucked up.
+			color = "[COLOR_RED]"
+			investigate_log("Reactor taking damage from the lack of coolant", INVESTIGATE_SINGULO)
+	//Now, heat up the output and set our pressure.
+	coolant_output.temperature = CELSIUS_TO_KELVIN(temperature) //Heat the coolant output gas that we just had pass through us.
```
Comment (redmoogle) [inline code/modules/power/reactor/rbmk.dm]:
```suggestion
	coolant_output.set_temperature(CELSIUS_TO_KELVIN(temperature)
```

**Comment by warface1234455** [comment]
pog pog

File: code/modules/power/reactor/rbmk.dm [hunk 5]
```diff
@@ -0,0 +1,1031 @@
...
+	var/radioactivity_spice_multiplier = 1 //Some gasses make the reactor a bit spicy.
+	var/depletion_modifier = 0.035 //How rapidly do your rods decay
+	gas_absorption_effectiveness = gas_absorption_constant
+	//Next up, handle moderators!
+//	if(moderator_input.total_moles() >= minimum_coolant_level)
+	var/total_fuel_moles = 0
+	moderator_input.assert_gas(/datum/gas/plasma)
+	moderator_input.assert_gas(/datum/gas/tritium)
+	total_fuel_moles += moderator_input.gases[/datum/gas/plasma][MOLES]
+	total_fuel_moles += moderator_input.gases[/datum/gas/tritium][MOLES]*10
```
Comment (redmoogle) [inline code/modules/power/reactor/rbmk.dm]:
```suggestion
	total_fuel_moles += moderator_input.get_moles(/datum/gas/plasma)
	total_fuel_moles += moderator_input.get_moles(/datum/gas/tritium)
```

File: code/modules/power/reactor/rbmk.dm [hunk 6]
```diff
@@ -0,0 +1,1031 @@
...
+		if(istype(I, /obj/item/reagent_containers/food) && !istype(I, /obj/item/reagent_containers/food/drinks))
+			playsound(src, pick('sound/machines/fryer/deep_fryer_1.ogg', 'sound/machines/fryer/deep_fryer_2.ogg'), 100, TRUE)
+			var/obj/item/reagent_containers/food/grilled_item = I
+			if(prob(80))
+				return //To give the illusion that it's actually cooking omegalul.
+			switch(power)
+				if(20 to 39)
+					grilled_item.name = "grilled [initial(grilled_item.name)]"
+					grilled_item.desc = "[initial(I.desc)] It's been grilled over a nuclear reactor."
+					if(!grilled_item.foodtype & TRAIT_FOOD_GRILLED)
```
Comment (redmoogle) [inline code/modules/power/reactor/rbmk.dm]:
```suggestion
					if(!(grilled_item.foodtype & TRAIT_FOOD_GRILLED))
```

**Comment by ghost** [comment]
It is fucked

File: code/modules/power/reactor/rbmk.dm [hunk 7]
```diff
@@ -0,0 +1,1031 @@
...
+						grilled_item.foodtype |= TRAIT_FOOD_GRILLED
+				if(40 to 70)
+					grilled_item.name = "heavily grilled [initial(grilled_item.name)]"
+					grilled_item.desc = "[initial(I.desc)] It's been heavily grilled through the magic of nuclear fission."
+					if(!grilled_item.foodtype & TRAIT_FOOD_GRILLED)
```
Comment (redmoogle) [inline code/modules/power/reactor/rbmk.dm]:
```suggestion
					if(!(grilled_item.foodtype & TRAIT_FOOD_GRILLED)
```

File: code/modules/power/reactor/rbmk.dm [hunk 8]
```diff
@@ -0,0 +1,1031 @@
...
+						grilled_item.foodtype |= TRAIT_FOOD_GRILLED
+				if(70 to 95)
+					grilled_item.name = "Three-Mile Nuclear-Grilled [initial(grilled_item.name)]"
+					grilled_item.desc = "A [initial(grilled_item.name)]. It's been put on top of a nuclear reactor running at extreme power by some badass engineer."
+					if(!grilled_item.foodtype & TRAIT_FOOD_GRILLED)
```
Comment (redmoogle) [inline code/modules/power/reactor/rbmk.dm]:
```suggestion
					if(!(grilled_item.foodtype & TRAIT_FOOD_GRILLED)
```

File: code/modules/power/reactor/rbmk.dm [hunk 9]
```diff
@@ -0,0 +1,1031 @@
...
+						grilled_item.foodtype |= TRAIT_FOOD_GRILLED
+				if(95 to INFINITY)
+					grilled_item.name = "Ultimate Meltdown Grilled [initial(grilled_item.name)]"
+					grilled_item.desc = "A [initial(grilled_item.name)]. A grill this perfect is a rare technique only known by a few engineers who know how to perform a 'controlled' meltdown whilst also having the time to throw food on a reactor. I'll bet it tastes amazing."
+					if(!grilled_item.foodtype & TRAIT_FOOD_GRILLED)
```
Comment (redmoogle) [inline code/modules/power/reactor/rbmk.dm]:
```suggestion
					if(!(grilled_item.foodtype & TRAIT_FOOD_GRILLED)
```

File: code/modules/power/reactor/rbmk.dm [hunk 10]
```diff
@@ -0,0 +1,1031 @@
...
+		icon_state = "inferior"
+		AddComponent(/datum/component/radioactive, 1500 , src)
+
+	else
+		fuel_power = 0.10
+
+/obj/item/fuel_rod/Initialize()
+	.=..()
+	AddComponent(/datum/component/radioactive, 350 , src)
+	AddComponent(/datum/component/two_handed, require_twohands=TRUE)
```
Comment (redmoogle) [inline code/modules/power/reactor/rbmk.dm:587]: We only have `/obj/item/twohanded` iirc

**Review by redmoogle — CHANGES_REQUESTED** [review]

File: code/modules/power/reactor/rbmk.dm (hunk 10 above)
Comment (ghost) [reply code/modules/power/reactor/rbmk.dm:587]: i understood

File: code/modules/power/reactor/rbmk.dm [hunk 11]
```diff
@@ -248,7 +248,7 @@ The reactor CHEWS through moderator. It does not do this slowly. Be very careful
 			color = "[COLOR_RED]"
 			investigate_log("Reactor taking damage from the lack of coolant", INVESTIGATE_SINGULO)
 	//Now, heat up the output and set our pressure.
-	coolant_output.temperature = CELSIUS_TO_KELVIN(temperature) //Heat the coolant output gas that we just had pass through us.
+	coolant_output.set_temperature(CELSIUS_TO_KELVIN(temperature)
```
Comment (redmoogle) [inline code/modules/power/reactor/rbmk.dm]: I forgot the closing bracket :skull:
```suggestion
	coolant_output.set_temperature(CELSIUS_TO_KELVIN(temperature))
```

File: code/modules/power/reactor/rbmk.dm [hunk 12]
```diff
@@ -0,0 +1,1029 @@
...
+						grilled_item.foodtype |= TRAIT_FOOD_GRILLED
+				if(70 to 95)
+					grilled_item.name = "Three-Mile Nuclear-Grilled [initial(grilled_item.name)]"
+					grilled_item.desc = "A [initial(grilled_item.name)]. It's been put on top of a nuclear reactor running at extreme power by some badass engineer."
+					if(!(grilled_item.foodtype & TRAIT_FOOD_GRILLED)
+						grilled_item.foodtype |= TRAIT_FOOD_GRILLED
+				if(95 to INFINITY)
+					grilled_item.name = "Ultimate Meltdown Grilled [initial(grilled_item.name)]"
+					grilled_item.desc = "A [initial(grilled_item.name)]. A grill this perfect is a rare technique only known by a few engineers who know how to perform a 'controlled' meltdown whilst also having the time to throw food on a reactor. I'll bet it tastes amazing."
+					if(!(grilled_item.foodtype & TRAIT_FOOD_GRILLED)
```
Comment (redmoogle) [inline code/modules/power/reactor/rbmk.dm]:
```suggestion
					if(!(grilled_item.foodtype & TRAIT_FOOD_GRILLED))
```
I think im retarded

File: code/modules/power/reactor/rbmk.dm [hunk 13]
```diff
@@ -0,0 +1,1029 @@
...
+						grilled_item.foodtype |= TRAIT_FOOD_GRILLED
+				if(40 to 70)
+					grilled_item.name = "heavily grilled [initial(grilled_item.name)]"
+					grilled_item.desc = "[initial(I.desc)] It's been heavily grilled through the magic of nuclear fission."
+					if(!(grilled_item.foodtype & TRAIT_FOOD_GRILLED)
+						grilled_item.foodtype |= TRAIT_FOOD_GRILLED
+				if(70 to 95)
+					grilled_item.name = "Three-Mile Nuclear-Grilled [initial(grilled_item.name)]"
+					grilled_item.desc = "A [initial(grilled_item.name)]. It's been put on top of a nuclear reactor running at extreme power by some badass engineer."
+					if(!(grilled_item.foodtype & TRAIT_FOOD_GRILLED)
```
Comment (redmoogle) [inline code/modules/power/reactor/rbmk.dm]:
```suggestion
					if(!(grilled_item.foodtype & TRAIT_FOOD_GRILLED))
```

File: code/modules/power/reactor/rbmk.dm [hunk 14]
```diff
@@ -0,0 +1,1029 @@
...
+			switch(power)
+				if(20 to 39)
+					grilled_item.name = "grilled [initial(grilled_item.name)]"
+					grilled_item.desc = "[initial(I.desc)] It's been grilled over a nuclear reactor."
+					if(!(grilled_item.foodtype & TRAIT_FOOD_GRILLED))
+						grilled_item.foodtype |= TRAIT_FOOD_GRILLED
+				if(40 to 70)
+					grilled_item.name = "heavily grilled [initial(grilled_item.name)]"
+					grilled_item.desc = "[initial(I.desc)] It's been heavily grilled through the magic of nuclear fission."
+					if(!(grilled_item.foodtype & TRAIT_FOOD_GRILLED)
```
Comment (redmoogle) [inline code/modules/power/reactor/rbmk.dm]:
```suggestion
					if(!(grilled_item.foodtype & TRAIT_FOOD_GRILLED))
```

Comment (ghost) [reply code/modules/power/reactor/rbmk.dm]: Already done

**Comment by maxion12345** [comment]
good luck lmao I've tried this twice
also don't forget the part that stops rads if items are in pools

Mainly I can just never get the reactor to be stable (at least for any meaningful period of time and generating meaningful power) so hopefully you'll have success where I failed 

Once it gets to a stable point you might consider replacing the alarm with radio calls similar to the sm. Also blowout make fallout go brr.

**Comment by ghost** [comment]
> Mainly I can just never get the reactor to be stable (at least for any meaningful period of time and generating meaningful power) so hopefully you'll have success where I failed

It seems that i wouldn't sucess here

**Comment by maxion12345** [comment]
> > Mainly I can just never get the reactor to be stable (at least for any meaningful period of time and generating meaningful power) so hopefully you'll have success where I failed
> 
> It seems that i wouldn't sucess here

Yeah I don't know how nsv does it. The blowout mechanic is really neat. Might take a crack at their stormdrive instead and see if that's somehow more stable. Of course, one of the big issues is that NSV's reactors typically take fuel directly from their Atmos whereas most of our engines are self-contained so the Atmos techs can make autism furnaces and churn out 3gazilloon metallic hydrogen. 

Definitely worth looking into though

**Pull request closed without merging**